alacritty/src/renderer/rect.rs

```diff
@@ -0,0 +1,259 @@
+            let index = index as u16 * 4;
+            indices.push(index);
+            indices.push(index + 1);
+            indices.push(index + 2);
+            indices.push(index + 2);
+            indices.push(index + 3);
+            indices.push(index + 1);
+        }
+
+        unsafe {
+            // Allocate buffers.
+            gl::GenVertexArrays(1, &mut vao);
+            gl::GenBuffers(1, &mut vbo);
+            gl::GenBuffers(1, &mut ebo);
+
+            // Bind EBO to VAO.
+            gl::BindVertexArray(vao);
+            gl::BindBuffer(gl::ELEMENT_ARRAY_BUFFER, ebo);
+
+            // Pre-upload index buffer for max possible rects.
+            gl::BufferData(
+                gl::ELEMENT_ARRAY_BUFFER,
+                (indices.len() * std::mem::size_of::<u16>()) as isize,
+                indices.as_ptr() as *const _,
+                gl::STATIC_DRAW,
+            );
+
+            // VBO binding is not part ot VAO, but attributes are.
+            gl::BindBuffer(gl::ARRAY_BUFFER, vbo);
+
+            let mut index = 0;
+            let mut size = 0;
+
+            macro_rules! add_attr {
+                ($count:expr, $gl_type:expr, $normalize:expr, $type:ty) => {
+                    gl::VertexAttribPointer(
+                        index,
+                        $count,
+                        $gl_type,
+                        $normalize,
+                        size_of::<Vertex>() as i32,
+                        size as *const _,
+                    );
+                    gl::EnableVertexAttribArray(index);
+
+                    #[allow(unused_assignments)]
+                    {
+                        size += $count * size_of::<$type>();
+                        index += 1;
+                    }
+                };
+            }
+
+            // Position.
+            add_attr!(2, gl::FLOAT, gl::FALSE, f32);
+
+            // Color.
+            add_attr!(4, gl::UNSIGNED_BYTE, gl::TRUE, u8);
+
+            // Reset buffer bindings.
+            gl::BindVertexArray(0);
+            gl::BindBuffer(gl::ARRAY_BUFFER, 0);
+        }
+
+        Ok(Self { vao, vbo, ebo, program, vertices: Vec::new() })
+    }
+
+    pub fn draw(&mut self, size_info: &SizeInfo, rects: Vec<RenderRect>) {
+        // Setup bindings. VAO will set up attribs and EBO, but not VBO.
+        unsafe {
+            gl::BindVertexArray(self.vao);
+
+            // Bind VBO only once for buffer data upload in draw_accumulated.
+            gl::BindBuffer(gl::ARRAY_BUFFER, self.vbo);
+
+            gl::UseProgram(self.program.id);
+        }
+
+        let center_x = size_info.width() / 2.;
+        let center_y = size_info.height() / 2.;
+
+        for rect in &rects {
+            self.append_rect(center_x, center_y, rect);
+            if self.vertices.len() == MAX_VERTICES {
+                self.draw_accumulated();
+            }
+        }
+
+        if !self.vertices.is_empty() {
+            self.draw_accumulated();
+        }
```

Comment:
Use https://doc.rust-lang.org/std/slice/struct.Chunks.html and invoke closure that appends & draws for each chunk.

Reply:
Yeah, I was thinking something like that. Though taking chunks might make calling `append_rect` a tiny bit more annoying. It would probably still be better than the current version though.

Reply:
Chunks is a nice suggestion, I didn't know it existed. Thanks! I'll try to apply it.